content/blog/pointers-and-references-cpp/index.md

```diff
@@ -0,0 +1,153 @@
+
+Here's an example:
+
+```cpp
+	// Defines int variable num
+	int num = 12;
+
+	// Declares pointer p
+	// Ignore the '&' symbol here, we'll discuss that next
+	int *p = &num;
+
+	// Prints out the memory address that p is pointing to
+	cout << p << endl;
+
+	// Prints out the memory address of num
+	// Again ignore the '&' symbol here, we'll discuss that next
+	cout << &num << endl;
+
+	// Dereferences the pointer p
+	cout << *p << endl;
+
+	// Prints out the value num
+	cout << num << endl;
+```
+  
+This should print out something like...
+
+```
+	0xffffcc14
+	0xffffcc14
+	12
+	12
+```
+
+As you can see, the pointer p holds the memory address of num, and when you use the dereference operator, the value at num is printed out.
+
+Pointers can also get a lot more complex and must be used in certain situations. For example, if you put an object on the heap (Check out my article on Virtual Memory to learn more about heap memory) then you will have to use a pointer because you can't access the heap directly. So, instead of having a pointer to an address on the stack, it will point to an address on the heap. You might even find yourself using double or triple pointers as you get more used to them.
+
+
+## What is a Reference? {#what-is-a-reference}
+
+In simple terms, a reference is simply the address of whatever you're passing. The difference between a pointer and a reference lies in the fact that a reference **address** where a value of some sort is being stored, a pointer has a value at an address different from the one it’s pointing to. I like to consider the **&** operator the "reference operator" even though I'm pretty sure that's not actually what it is called. I used this operator in the last example, and it's pretty straightforward.
+
+```cpp  
+	int num = 12;
+	int *val = &num;
+	int **doublePointer = &val;
+
+	cout << "Address of num: " << &num << endl;
+	cout << "Value of val" << val << endl;
+	
+	// Pointers have memory addresses too!
+	cout << "Address of val" << &val << endl;
+	
+	// Prints out every phase of the double pointer
+	cout << &doublePointer << " : " << doublePointer << " : ";
+
+	cout << *doublePointer << " : " << **doublePointer << endl;
+```
+
+The output will look something like this...
+
+```
+	Address of num: 0xffffcc1c
+	Value of val: 0xffffcc1c
+	Address of val: 0xffffcc10
+	0xffffcc08 : 0xffffcc10 : 0xffffcc1c : 12
+```
+  
+As you can see, all the **&** operator does is gives you the memory address at its specific spot in memory. I also included a small example of a double-pointer which is just one more layer of abstraction then another pointer. You can see how the memory addresses line up.
+
+Here’s what this looks like in memory with more easily understandable addresses in a “0x…” format.
+
+**![Memory Example](./memory.png)**
```

Comment:
This should give an image with an Alt of `Memory Example` and no title. Does the formatting MD engine allow titles?

Reply:
Remark should allow titles as well as alt code for images